internal/terraform/node_resource_abstract_instance.go

```diff
@@ -2059,44 +2066,38 @@ func (n *NodeAbstractResourceInstance) evalDestroyProvisionerConfig(ctx EvalCont
 }
 
 // apply accepts an applyConfig, instead of using n.Config, so destroy plans can
-// send a nil config. Most of the errors generated in apply are returned as
-// diagnostics, but if provider.ApplyResourceChange itself fails, that error is
-// returned as an error and nil diags are returned.
+// send a nil config. The keyData information can be empty if the config is
+// nil, since it is only used to evaluate the configuration.
 func (n *NodeAbstractResourceInstance) apply(
 	ctx EvalContext,
 	state *states.ResourceInstanceObject,
 	change *plans.ResourceInstanceChange,
 	applyConfig *configs.Resource,
-	createBeforeDestroy bool) (*states.ResourceInstanceObject, instances.RepetitionData, tfdiags.Diagnostics) {
+	keyData instances.RepetitionData,
```

Comment:
Technically anything which has access to the `EvalContext` can get the "repetition data" like this:

```go
    exp := ctx.InstanceExpander()
    keyData := exp.GetResourceInstanceRepetitionData(n.Addr())
```

In the original design for this "instance expander" thing it was intended to act as the source of record for all of this expansion-related information specifically so that we wouldn't need to explicitly pass all of this contextual data around.

I do feel in two minds about it because it feels like a violation of [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter), but we already widely use `ctx` as a big bag of shared "global" context, including in this very function where it uses `ctx` to do various other things that aren't strictly this function's business to do.

Reply:
Yes, I left this as close to the current code as possible. I did forget that the rep data was directly available though, because the existing apply code already re-evaluated it anyway. Switching out that code however runs into a no `expansion has been registered` panic, so I think there's something else going on here too.

Reply:
Hmm yes I suppose the fact that we've been passing it around through arguments meant that there was not previously any need to actually register it at the source of record. Unfortunate but not unexpected.

Hopefully in a future change we can find a suitable central point to evaluate this just once and register it in the expander, so it's clearer exactly whose responsibility it is to evaluate the `for_each` or `count` expressions.